[SPARK-12705] [SPARK-10777] [SQL] Analyzer Rule ResolveSortReferences

gatorsmile opened the pull request:
JIRA: https://issues.apache.org/jira/browse/SPARK-12705

**Scope:**
This PR is a general fix for sorting reference resolution when the child's `outputSet` does not have the order-by attributes (called, _missing attributes_):
- UnaryNode support is limited to `Project`, `Window`, `Aggregate`, `Distinct`, `Filter`, `RepartitionByExpression`.
- We will not try to resolve the missing references inside a subquery, unless the outputSet of this subquery contains it.

**General Reference Resolution Rules:**
- Jump over the nodes with the following types: `Distinct`, `Filter`, `RepartitionByExpression`. Do not need to add missing attributes. The reason is their `outputSet` is decided by their `inputSet`, which is the `outputSet` of their children.
- Group-by expressions in `Aggregate`: missing order-by attributes are not allowed to be added into group-by expressions since it will change the query result. Thus, in RDBMS, it is not allowed. 
- Aggregate expressions in `Aggregate`: if the group-by expressions in `Aggregate` contains the missing attributes but aggregate expressions do not have it, just add them into the aggregate expressions. This can resolve the analysisExceptions thrown by the three TCPDS queries. 
- `Project` and `Window` are special. We just need to add the missing attributes to their `projectList`. 

**Implementation:**
1. Traverse the whole tree in a pre-order manner to find all the resolvable missing order-by attributes. 
2. Traverse the whole tree in a post-order manner to add the found missing order-by attributes to the node if their `inputSet` contains the attributes. 
3. If the origins of the missing order-by attributes are different nodes, each pass only resolves the missing attributes that are from the same node. 

**Risk:** 
Low. This rule will be trigger iff `!s.resolved && child.resolved` is true. Thus, very few cases are affected.

gatorsmile committed:
window function: Sorting columns are not in Project

gatorsmile committed:
style fix.

SparkQA commented:
**[Test build #49052 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49052/consoleFull)** for PR 10678 at commit [`5ca4630`](https://github.com/apache/spark/commit/5ca463035bc6eaebd15e7cf332faeea157e5593e).

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -523,14 +523,37 @@ class Analyzer(
       case s @ Sort(ordering, global, p @ Project(projectList, child))
           if !s.resolved && p.resolved =>
-        val (newOrdering, missing) = resolveAndFindMissing(ordering, p, child)
+        val (newOrdering, missing, newChild): (Seq[SortOrder], Seq[Attribute], LogicalPlan) =
```
can you add some comment here about why we need two separate cases.

gatorsmile replied:
Sure, will do

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -523,14 +523,37 @@ class Analyzer(
-        val (newOrdering, missing) = resolveAndFindMissing(ordering, p, child)
+        val (newOrdering, missing, newChild): (Seq[SortOrder], Seq[Attribute], LogicalPlan) =
+          child match {
+            case Project(
```
btw i think this is way too complicated. you can simplify it if you don't specify all the fields, e.g.
```scala
case p1 @ Project(_, w @ Window(_, _, _, _, p2: Project)) =>
```

rxin replied:
and it also becomes more clear later which operator is each variable/expression for because you will need to specify "p1.projectList", "w.projectList", etc.

gatorsmile replied:
yeah, thank you! Will change it.

SparkQA commented:
**[Test build #49052 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49052/consoleFull)** for PR 10678 at commit [`5ca4630`](https://github.com/apache/spark/commit/5ca463035bc6eaebd15e7cf332faeea157e5593e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49052/
Test PASSed.

gatorsmile committed:
code cleaning and address comments.

SparkQA commented:
**[Test build #49054 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49054/consoleFull)** for PR 10678 at commit [`da6baf2`](https://github.com/apache/spark/commit/da6baf25488767ce6e73538b03f9195bba92b84e).

SparkQA commented:
**[Test build #49054 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49054/consoleFull)** for PR 10678 at commit [`da6baf2`](https://github.com/apache/spark/commit/da6baf25488767ce6e73538b03f9195bba92b84e).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49054/
Test FAILed.

gatorsmile committed:
Merge remote-tracking branch 'upstream/master' into sortWindows

gatorsmile commented:
retest this please.

SparkQA commented:
**[Test build #49058 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49058/consoleFull)** for PR 10678 at commit [`da6baf2`](https://github.com/apache/spark/commit/da6baf25488767ce6e73538b03f9195bba92b84e).

SparkQA commented:
**[Test build #49058 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49058/consoleFull)** for PR 10678 at commit [`da6baf2`](https://github.com/apache/spark/commit/da6baf25488767ce6e73538b03f9195bba92b84e).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49058/
Test FAILed.

gatorsmile commented:
retest this please.

SparkQA commented:
**[Test build #49060 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49060/consoleFull)** for PR 10678 at commit [`da6baf2`](https://github.com/apache/spark/commit/da6baf25488767ce6e73538b03f9195bba92b84e).

SparkQA commented:
**[Test build #49061 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49061/consoleFull)** for PR 10678 at commit [`b5de079`](https://github.com/apache/spark/commit/b5de0799650a86b8479eb053d7e3e65b23e5d75b).

SparkQA commented:
**[Test build #49060 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49060/consoleFull)** for PR 10678 at commit [`da6baf2`](https://github.com/apache/spark/commit/da6baf25488767ce6e73538b03f9195bba92b84e).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49060/
Test FAILed.

SparkQA commented:
**[Test build #49061 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49061/consoleFull)** for PR 10678 at commit [`b5de079`](https://github.com/apache/spark/commit/b5de0799650a86b8479eb053d7e3e65b23e5d75b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49061/
Test PASSed.

gatorsmile commented:
@davies Could you take a look if the fix covers the analysis resolution issue in TPCD? Thank you!

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -521,16 +521,31 @@ class Analyzer(
+          p.child match {
+            // Case 1: when WINDOW functions are used in the SELECT clause.
+            //   Example: SELECT sum(col1) OVER() FROM table1 ORDER BY col2
+            case p1 @ Project(_, w @ Window(_, _, _, _, p2: Project)) =>
```
What if the child of Window is not a Project (a Aggregation or other)? Or there are multiple projections ?

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -521,16 +521,31 @@ class Analyzer(
           Project(p.output,
-            Sort(newOrdering, global,
-              Project(projectList ++ missing, child)))
+            Sort(newOrdering, s.global,
```
Should we continue to add the missing attributes here until it can be resolved?

gatorsmile replied:
This is a good suggestion!

davies commented:
@gatorsmile Thanks for working on this, but after this patch, TPCDS Q12/Q20/Q98 are still can't be resolved: " cannot resolve 'i_item_id' given input columns: [i_class, itemrevenue, i_category, i_current_price".

You can see all these queries here: https://github.com/databricks/spark-sql-perf/blob/master/src/main/scala/com/databricks/spark/sql/perf/tpcds/TPCDS_1_4_Queries.scala#L612

gatorsmile commented:
Let me do more investigation tomorrow. @davies : )

gatorsmile commented:
I can reproduce the problem using a simple query now:
```
select area, rank() over (partition by area order by month) as c1
from windowData group by product, area, month order by product
```

The logical plan is like 
```
'Sort ['p ASC], true
+- Project [area#1,c1#11]
   +- Project [area#1,month#0,tmp1#10,_we0#17,(_we0#17 + tmp1#10) AS c1#11]
      +- Window [area#1,month#0,tmp1#10], [rank(month#0) windowspecdefinition(area#1,month#0 ASC,ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS _we0#17], [area#1], [month#0 ASC]
         +- Project [area#1,month#0,tmp1#10]
            +- Subquery tmp
               +- Project [month#0,area#1,product#2 AS p#9,1 AS tmp1#10]
                  +- Subquery windowdata
                     +- LogicalRDD [month#0,area#1,product#2], MapPartitionsRDD[1] at apply at Transformer.scala:22
```

This does not match the existing two patterns. Thus, it is unable to resolve the sorting columns. I will try to find and write a general rule to handle the cases, as you suggested.

gatorsmile commented:
I found the sorting columns could be partially resolved by different nodes. For example, 
```select a.c1, b.c2 from t1 a, t1 b order by b.c3, a.c3```

I need extra time to write the fix. It becomes more complex than what I thought.

gatorsmile commented:
Like the example I posted above, it works in RDBMS. However, supporting it in Spark SQL is not trivial. In this example, `Join` operator is involved but it does not have a `projectList` attribute. Thus, we are unable to directly add missing attributes to it. Now, only two operators `Project` and `Window` are doable.  That means, we might see more JIRAs to complain it. 

In the long term, maybe we can add `projectList` to all the logical operators? That indicates we integrate `Project` into all the logical operators. 

I will check if the three TPCDS queries can be resolved without major code changes. Thanks!

davies commented:
@gatorsmile There is a JIRA to have projectList in Join, we can do that later. For this PR, we may just push the missing attributes until any JOIN.

gatorsmile commented:
Thank you! @davies Will upload the new version tonight.

gatorsmile committed:
address comments.

gatorsmile commented:
In this update, the code changes are trying to extend the existing support for the following scenario: ```sort by attributes that are not present in the SELECT clause```

Now, between the operator that can resolve the sorting columns and the top `Sort` operator, it can have one or multiple different operators of `Aggregate`, `Project`, `Subquery` and `Window`. The reason why we just support these four types is that only these four types we can add sorting attributes into their project list. In the future, we can easily extend the solution by supporting more types.

I am not sure if we should change the collection we used in this tail recursion function? Now, it is `Seq`. If the plan tree is huge, the performance could be not good. Thanks!

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49296/
Test FAILed.

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -73,6 +74,7 @@ class Analyzer(
       ResolveGroupingAnalytics ::
       ResolvePivot ::
       ResolveUpCast ::
+      ResolveAggregateFunctions ::
```
Why we need to change the order?

gatorsmile replied:
In this rule `ResolveAggregateFunctions`, it has a case for pushing down the expressions in `Sort` operator to the underlying `Aggregate`. This needs to be done earlier than the rule `ResolveSortReferences`; Otherwise, we might add unnecessary missing sorting attributes.

cloud-fan replied:
depending on rule orders of one batch is not a good idea. We should either add more `if` conditions to only capture needed cases, or separate them into multiple batches.

gatorsmile replied:
Yeah, you are right. Will do the change to exclude the cases that can be resolved by `ResolveAggregateFunctions`. Thanks!

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -521,38 +522,81 @@ class Analyzer(
           s // Nothing we can do here. Return original plan.
         }
+        else {
+          val missingPlanSet = missingPlans.toSet
```
I think we don't need this missingPlans, could be done like this:
```
var stop = false
val newChild = child transform {
 case Projection(list, child) if missingAttr.nonEmpty && !stop =>
    val newList = list ++ missingAttr
    missingAttr = missingAttr.filterNot(child.outputSet.contains)
    Projection(newList, child)
case j: Join => 
   stop = true
   j
}

```

gatorsmile replied:
Thank you! I will change the implementation. @davies 

Originally, I thought we were not allowed to use a `var` flag since all the rules in optimizer and analyzer do not use it. Now, if we can do it like that, it is much simpler to do the tree traversal. : )

gatorsmile committed:
address comments.

gatorsmile commented:
Based on the comments of @davies , this update is for avoiding the usage of extra buffer in the recursion function. It should perform faster and make codes simpler.

SparkQA commented:
**[Test build #49307 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49307/consoleFull)** for PR 10678 at commit [`27fcaa5`](https://github.com/apache/spark/commit/27fcaa5ad6a3b4228ef4fc46b963c1e818d2f5c4).

SparkQA commented:
**[Test build #49307 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49307/consoleFull)** for PR 10678 at commit [`27fcaa5`](https://github.com/apache/spark/commit/27fcaa5ad6a3b4228ef4fc46b963c1e818d2f5c4).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49307/
Test FAILed.

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -521,38 +522,90 @@ class Analyzer(
+              a.copy(groupingExpressions = newGroupExpressions,
+                aggregateExpressions = newAggregateExpressions)
+            case s: Subquery if !stop && missingAttrs.nonEmpty => s
+            case o =>
```
How about Filter (having clause)?

davies commented:
@gatorsmile For Subquery/Filter/Join, because they will have the same output as child/children, we should continue to push the attributes through them.

gatorsmile commented:
Yeah! Then, following the same rule, we can handle all the operators. : )

gatorsmile commented:
In the implementation, I think we need to use the extra collection to record which child can resolve the missing attributes when handling the binary operators. For example, `Join` has two children, we have to know which child should add missing Attributes (except left-semi and left-anti) . That means, we need to use the solution I did at the beginning. Is that OK to you?

yhuai commented:
Can you do do a quick search in the jira? There is a old jira for this issue. Can you mark that as a duplicated one? Thanks.

gatorsmile commented:
Sure, will do. @yhuai

gatorsmile commented:
Tomorrow is the deadline of one paper. Sorry for the delay, will update this PR tonight or tomorrow night. Thanks!

rxin commented:
(not my business - but out of curiosity, which conference is this?)

gatorsmile commented:
Sigmod Demo, but we just start writing it this week. I guess it will be rejected, but the topic is very interesting, I think. It is also a Spark related project. 

RDBMS transaction log captures the change history of the RDBMS transactional data. If available on big data platforms, the decoded transaction log opens up various new business opportunities. Highly scalable data replication from RDBMS transaction log to the big data platforms remains a challenge, and further research is required to enable efficient time travel and data retrieval for further analysis and analytics exploration.

gatorsmile committed:
added a support to more operators.

gatorsmile commented:
In this update: Added support for all the `UnaryNode` operators and `Join`. 
Todo: 
  - Will add more test cases to AnalysisSuite.
  - Will add a support for `Alias`.

SparkQA commented:
**[Test build #49539 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49539/consoleFull)** for PR 10678 at commit [`7fc98e4`](https://github.com/apache/spark/commit/7fc98e49a26fd03f398b2241b4cfd19e969b770e).

SparkQA commented:
**[Test build #49539 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49539/consoleFull)** for PR 10678 at commit [`7fc98e4`](https://github.com/apache/spark/commit/7fc98e49a26fd03f398b2241b4cfd19e969b770e).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49539/
Test FAILed.

gatorsmile committed:
Merge remote-tracking branch 'upstream/master' into sortWindows

gatorsmile committed:
style fix.

SparkQA commented:
**[Test build #49541 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49541/consoleFull)** for PR 10678 at commit [`522626b`](https://github.com/apache/spark/commit/522626bbd483054f441d2ca49bc06512901258ea).

SparkQA commented:
**[Test build #49541 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49541/consoleFull)** for PR 10678 at commit [`522626b`](https://github.com/apache/spark/commit/522626bbd483054f441d2ca49bc06512901258ea).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49541/
Test FAILed.

gatorsmile committed:
fixed the test case that might fail sometimes due to the sorted values are duplicate

SparkQA commented:
**[Test build #49545 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49545/consoleFull)** for PR 10678 at commit [`26945fa`](https://github.com/apache/spark/commit/26945fa63809a8671461404eb2e661e1605dc196).

SparkQA commented:
**[Test build #49545 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49545/consoleFull)** for PR 10678 at commit [`26945fa`](https://github.com/apache/spark/commit/26945fa63809a8671461404eb2e661e1605dc196).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49545/
Test PASSed.

gatorsmile committed:
added test cases

SparkQA commented:
**[Test build #49558 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49558/consoleFull)** for PR 10678 at commit [`bd3ed13`](https://github.com/apache/spark/commit/bd3ed13b9e78d59274cda6c243acc5e704bb2821).

gatorsmile commented:
The current commit is using a multi-pass reference resolution if the origins of the missing attributes are different nodes. Tonight, will try to reduce it to one pass.

davies commented:
@gatorsmile Since it will not have thousands of nodes, it's find to do multiple passes. I'd pick the simpler solution.

SparkQA commented:
**[Test build #49558 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49558/consoleFull)** for PR 10678 at commit [`bd3ed13`](https://github.com/apache/spark/commit/bd3ed13b9e78d59274cda6c243acc5e704bb2821).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49558/
Test PASSed.

gatorsmile commented:
Sure, will fix the bug and let you know when everything is done. Thanks! @davies

gatorsmile committed:
fixed bugs.

gatorsmile commented:
In this update, the fix is ready to review:
  - Group-by expressions in `Aggregate`: missing order-by columns are not allowed to be added into group-by expressions since it will change the query result. 
  - UnaryNode support is limited to `Project`, `Window`, `Aggregate`, `Distinct`, `Filter`, `Limit`, `RedistributeData`, `Repartition`, `RepartitionByExpression`, `Sample`, `Sort`, `SortPartitions`, `Subquery`, `With`, and `WithWindowDefinition`. 
  - BinaryNode support is still limited to `Join`. 

The basic idea of algorithms is simple. 
  1. Traverse the whole tree in a pre-order manner to find all the resolvable missing order-by attributes. 
  2. Traverse the whole tree in a post-order manner to add the found missing order-by attributes to the node if their `inputSet` contains the attributes. 

Let me know if any part is not appropriate. Thank you! @davies @rxin @yhuai

SparkQA commented:
**[Test build #49582 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49582/consoleFull)** for PR 10678 at commit [`831baf5`](https://github.com/apache/spark/commit/831baf515faae0f12fae0f8b50297c05292e9e16).

rxin commented:
@davies can you please hold off merging this until more people look at it? It is getting more and more complicated.

davies commented:
@gatorsmile Could you update the description of the PR (according to your last update)

davies commented:
Agreed that it's more complicated that expected, should pay more attention on it. If it's too complicated that we are not confident enough, may be we should narrow down the cases, only support the cases that are simple enough.

@gatorsmile If we can't go with this current complicated approach, I'm really sorry for that.

SparkQA commented:
**[Test build #49582 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49582/consoleFull)** for PR 10678 at commit [`831baf5`](https://github.com/apache/spark/commit/831baf515faae0f12fae0f8b50297c05292e9e16).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49582/
Test PASSed.

gatorsmile commented:
@davies That is fine. : ) I just want to help you make Spark better. Updated the PR descriptions. At the beginning, I tried to write simpler code, but I finally realized the problem is not simple. Actually, it can be more complicated if we also issue the user friendly error messages to explain why Analyzer cannot resolve the missing attributes, especially for the `Aggregate` operator.

@rxin Yeah. Agree! More reviews are needed for sure. Feel free to let me know if any change is needed.

gatorsmile commented:
Also cc @marmbrus @cloud-fan who touched the related codes before. Thank you!

yhuai commented:
@gatorsmile I thought only window functions still have this kind of problem. Is there any other case that our analyzer does not handle?

gatorsmile commented:
Based on what @davies described, TPCDS Q12/Q20/Q98 still can't be resolved. After investigation, the problem is caused by `Aggregate`:
  - Group-by expressions in Aggregate: missing order-by attributes are not allowed to be added into group-by expressions since it will change the query result. Thus, in RDBMS, it is not allowed.
  - Aggregate expressions in Aggregate: if the group-by expressions in Aggregate contains the missing attributes but aggregate expressions do not have it, just add them into the aggregate expressions. This can resolve the analysisExceptions thrown by the three TCPDS queries.

You can see all these queries here: https://github.com/databricks/spark-sql-perf/blob/master/src/main/scala/com/databricks/spark/sql/perf/tpcds/TPCDS_1_4_Queries.scala#L612

Another typical example is `Join`: ```select a.c1, b.c2 from t1 a, t1 b order by b.c3, a.c3```. 

After further discussion with @davies , `Filter` and the other UnaryNode operators we can jump over when resolving sorting attributes.

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala:
```diff
@@ -76,6 +76,118 @@ class AnalysisSuite extends AnalysisTest {
+      .sortBy('b.asc, 'c.desc)
+    val expected1 = testRelation2
+      .where(a.cast(DoubleType) > Literal(0).cast(DoubleType)).select(a, b, c)
+      .where(b.cast(DoubleType) > Literal(0).cast(DoubleType)).select(a, b, c)
```
We can pick the attributes carefully(like pick a int column instead of `a`) to avoid adding cast in analyzed plan, to make the test easier to read.

gatorsmile replied:
Sure, will do. Thanks!

gatorsmile committed:
updated test case.

gatorsmile committed:
updated test case.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49648/
Test FAILed.

gatorsmile commented:
retest this please.

SparkQA commented:
**[Test build #49666 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49666/consoleFull)** for PR 10678 at commit [`598a673`](https://github.com/apache/spark/commit/598a673d1b01ebef6a89582f8e63ab487e465e71).

SparkQA commented:
**[Test build #49666 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49666/consoleFull)** for PR 10678 at commit [`598a673`](https://github.com/apache/spark/commit/598a673d1b01ebef6a89582f8e63ab487e465e71).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49666/
Test PASSed.

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -521,38 +524,138 @@ class Analyzer(
+          logDebug(s"Failed to find $unresolvableAttrs in ${child.output.mkString(", ")}")
           s // Nothing we can do here. Return original plan.
         }
+        else {
```
`} else {`

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -521,38 +524,138 @@ class Analyzer(
+            case a: Aggregate =>
+              val newAggregateExpressions = a.aggregateExpressions ++
+                findNotResolvedMissingAttrs(
+                  a.aggregateExpressions, a.inputSet, missingResolvableAttrs)
```
Should this be `findNotResolvedMissingAttrs(a.groupingExpressions, a.inputSet, missingResolvableAttrs)`?

gatorsmile replied:
Here, we expect `a.aggregateExpressions` misses the order-by columns. Maybe, we can change it  to `findNotResolvedMissingAttrs(a.aggregateExpressions, a.groupingExpressions, missingResolvableAttrs)`

davies replied:
If we add a column into `aggregateExpressions`, but it's not part of `groupingExpressions`, how the Aggregate work?

gatorsmile replied:
I need to put a comment to explain `findNotResolvedMissingAttrs`. It looks confusing. 

`findNotResolvedMissingAttrs(a.aggregateExpressions, a.groupingExpressions, missingResolvableAttrs)`  will return all the attributes of `missingResolvableAttrs` that are in `a.groupingExpressions`, but if `a.aggregateExpressions` already has it, we will not add it again. 

The complexity is introduced due to the support of grandchildren of `Join`. Now, since we do not support this case, we can simplify it.

davies replied:
I see, agreed.

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -521,38 +524,138 @@ class Analyzer(
           s // Nothing we can do here. Return original plan.
         }
+        else {
+          // Transform the whole tree in post-order. Add into the self's outputSet
```
"Add the missing attributes into projectList of Project/Window or aggregateExpressions of Aggregate, if they are in the inputSet but not in the outputSet of the plan."

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -521,38 +524,138 @@ class Analyzer(
+        missingAttrs: Seq[Attribute]): Seq[Attribute] = {
+      val resolvedAttrs =
+        missingAttrs.filter(attr => inputSet.exists(_.semanticEquals(attr)))
+      resolvedAttrs.filterNot(attr => outputSet.exists(_.semanticEquals(attr)))
```
If missingAttrs is AttributeSet, this should be as easy as `missingAttrs.intersect(inputSet) -- outputSet`

gatorsmile replied:
If we do not resolve sort references whose sources are below `subquery`, we can do it in your way. Otherwise, `semanticEquals` has to be used. Thanks! Let me use the simplified way.

davies replied:
I see. Once we agreed that we don't need to support Join, we can do that.

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -521,38 +524,138 @@ class Analyzer(
+    private def collectResolvableMissingAttrs(
+        ordering: Seq[SortOrder],
+        plans: mutable.Queue[LogicalPlan]): (Seq[SortOrder], Seq[Attribute]) = {
+      if (plans.isEmpty) (Seq.empty[SortOrder], Seq.empty[Attribute])
```
Add `{}` for this case

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -521,38 +524,138 @@ class Analyzer(
+            }
+            // If missingAttrs is empty, do not traverse its child and just try the next
+            else collectResolvableMissingAttrs(ordering, plans)
+          case join @ Join(left, right, joinType, _) =>
```
I think we should stop at Join, if it can't be resolved by outputSet of Join, then we can't resolve it, because the output of Join should already include all the outputSet of the children (exclude semi join). The projection/aggregation below Join are part of subquery, we could not resolve the ordering from subqueries.

gatorsmile replied:
Ok, I see.

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -521,38 +524,138 @@ class Analyzer(
+              (newOrdering, missingAttrs)
+            }
+            // If missingAttrs is empty, do not traverse its child and just try the next
+            else collectResolvableMissingAttrs(ordering, plans)
```
I think we should stop here, the Project/Aggregate below this Aggregate should be part of a subquery.

gatorsmile replied:
Yeah, if we do not need to handle the grandchildren of Join, the whole solution is simplified a lot. We do not need these. Will send you a new version for reviews.

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -521,38 +524,138 @@ class Analyzer(
+              un.isInstanceOf[SortPartitions] ||
+              un.isInstanceOf[Subquery] ||
+              un.isInstanceOf[With] ||
+              un.isInstanceOf[WithWindowDefinition] =>
```
It's correct to list all these here, but I can't image a case that those except Filter are needed.

Subquery/With/WithWindowDefinition/Limit should be part of subquery, How to use use Distinct/RedistributeData/Repartition/RepartitionByExpression in SQL query? For DataFrame, we may not support adding missing attributes for Sort, it's a little bit confusing.

gatorsmile replied:
In SQL, `DISTRIBUTE BY` and `CLUSTER BY` will use `RepartitionByExpression`. `SELECT DISTINCT` will use `Distinct`

davies replied:
If we have `DISTRIBUTE/CLUSTER BY` and `ORDER BY `, which one will come first? In practice, they should not be used together.

`Distinct` sounds good.

davies commented:
@gatorsmile I took a pass, it's correct overall. My only concern is that it's too powerful that could introduce some side effects that we don't want. For example:
```
 df.select('a).limit(10).orderBy('b)
```
We may not want support these complicated use case. With DataFrame API, we could easily construct a plan as we want.

Without subquery, SQL query can only have one SELECT clause, it's reasonable to support that (at least other database support that), or you have to use subquery.

Here is the cases that we should support:
```
SELECT project_list FROM tables GROUP BY keys HAVING condition ORDER BY ordering LIMIT N
```

The columns used in HAVING and ORDER BY could be not presented in SELECT, we should still resolve them. 

Between Sort and Join/Scan, it could have Project/Filter/Window/Aggregate, no others.

davies commented:
cc @rxin

gatorsmile commented:
@davies How about these three `DISTRIBUTE BY`, `CLUSTER BY` and `SELECT DISTINCT`? Should we also support them?

davies commented:
DISTINCT sounds good, can we use DISTRIBUTE BY, CLUSTER BY together with ORDER BY? they all change the final order.

gatorsmile commented:
I saw a test case in `LogicalPlanToSQLSuite.scala`. 
```
  test("distribute by with sort by") {
    checkHiveQl("SELECT id FROM t0 DISTRIBUTE BY id SORT BY id")
  }
```
Do you think users could use it in this way?

davies commented:
@gatorsmile Let's keep that (DISTRIBUTE BY, CLUSTER BY) for now (even the query does not make sense to me).

gatorsmile commented:
Thank you for your reviews! @davies

gatorsmile commented:
Since we do not plan to support sort reference resolution inside subquery, should we just close the following JIRA? https://issues.apache.org/jira/browse/SPARK-10777

davies commented:
@gatorsmile I think that JIRA still valid, because the subquery already output the required attribute, we don't need to add the missing attributes into the project list inside subquery. I believe this PR could fix that.

davies commented:
Could you also include that in the title?

gatorsmile commented:
Sure, let me add it.

gatorsmile committed:
address comments.

SparkQA commented:
**[Test build #49765 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49765/consoleFull)** for PR 10678 at commit [`1964884`](https://github.com/apache/spark/commit/1964884afee035f1e13bd248073656ec43de2223).

SparkQA commented:
**[Test build #49765 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49765/consoleFull)** for PR 10678 at commit [`1964884`](https://github.com/apache/spark/commit/1964884afee035f1e13bd248073656ec43de2223).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49765/
Test PASSed.

gatorsmile commented:
@davies The code is ready to review, after addressing all the comments. 

@davies @cloud-fan The current implementation of `ResolveAggregateFunctions` also try to resolve the missing sort references, but it has a limit. The logics is triggered iff `Aggregate` is the child of `Sort`. I am just thinking if we should move it into the rule `ResolveSortReferences`?

marmbrus commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -521,38 +522,96 @@ class Analyzer(
+            case w: Window =>
+              w.copy(projectList = w.projectList ++
+                missingResolvableAttrs.filter((w.inputSet -- w.outputSet).contains))
+            case a: Aggregate =>
```
This case can never happen right?

gatorsmile replied:
In the following query, we can trigger this case. Actually, this query is based on the failed TPCDS queries. 
```select area, rank() over (partition by area order by month) as c1 from windowData group by product, area, month order by product, area```

If we remove this case, we will get this error:
```
Failed to analyze query: org.apache.spark.sql.AnalysisException: resolved attribute(s) product#2 missing from area#1,c1#39 in operator !Sort [product#2 ASC,area#1 ASC], true;
Project [area#1,c1#48]
+- !Sort [product#2 ASC,area#1 ASC], true
   +- Project [area#1,c1#48]
      +- Project [area#1,month#0,c1#48,c1#48]
         +- Window [area#1,month#0], [rank(month#0) windowspecdefinition(area#1,month#0 ASC,ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS c1#48], [area#1], [month#0 ASC]
            +- Aggregate [product#2,area#1,month#0], [area#1,month#0]
               +- Subquery windowdata
                  +- LogicalRDD [month#0,area#1,product#2], MapPartitionsRDD[1] at apply at Transformer.scala:22
```

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -521,38 +522,96 @@ class Analyzer(
-          if !s.resolved && p.resolved =>
-        val (newOrdering, missing) = resolveAndFindMissing(ordering, p, child)
+      // Here, this rule only resolves the missing sort references if the child is not Aggregate
+      //   Another rule ResolveAggregateFunctions will resolve that case.
```
Can we move the sort order resolution in `ResolveAggregateFunctions` to here too? Now it's hard to think about how these 2 rules interact...

gatorsmile replied:
Yeah, that is what I asked. Will do it. Thanks!

gatorsmile replied:
@cloud-fan I kept the function implementation in the `ResolveAggregateFunctions`, but I called the function in `ResolveSortReferences`. Since the rule `ResolveAggregateFunctions` covers two cases (`filter` and `sort`), I am afraid the code readers might feel confused if we split them into two rules. This function call is public. I am not sure if this way is appropriate?

davies replied:
The purpose of that rule is to resolve the aggregate functions (resolving expressions for Sort is kind of side effects), it's better have that in the `ResolveAggregateFunctions`. This also help to reduce the changes in this PR.

gatorsmile committed:
addressed comments.

SparkQA commented:
**[Test build #50449 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50449/consoleFull)** for PR 10678 at commit [`ba02f46`](https://github.com/apache/spark/commit/ba02f4695e4bfd07a9bef72f783bef3894d8191e).

SparkQA commented:
**[Test build #50449 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50449/consoleFull)** for PR 10678 at commit [`ba02f46`](https://github.com/apache/spark/commit/ba02f4695e4bfd07a9bef72f783bef3894d8191e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50449/
Test PASSed.

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -521,38 +522,99 @@ class Analyzer(
-      case s @ Sort(ordering, global, p @ Project(projectList, child))
-          if !s.resolved && p.resolved =>
-        val (newOrdering, missing) = resolveAndFindMissing(ordering, p, child)
+      case s @ Sort(_, _, a: Aggregate) if a.resolved =>
```
Do we need to check that `!s.resolved`, we have that in next case.

cloud-fan replied:
to be more clear, `ResolveSortReferences` tries to resolve attributes in `SortOrder`s, and `ResolveAggregateFunctions` tries to resolve aggregate functions in unexpected places(filter or sort), right?

So I think we should skip sort with aggregate functions here, i.e.
`case s: Sort if s.order.exists(ResolveAggregateFunctions.containsAggregate) => s`, and add comment to say this case should be handled in `ResolveAggregateFunctions`

davies replied:
There could be missing attributes together with aggregate functions, will that work?

gatorsmile replied:
@davies The missing attributes are also handled in `ResolveAggregateFunctions`. Thus it works. To answer your first question regarding `!s.resolved`, this is part of the algorithm design in the rule `ResolveAggregateFunctions`, as shown below: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L706-L708

gatorsmile replied:
@cloud-fan Sure, let me change it. Thanks!

cloud-fan replied:
`ResolveAggregateFunctions` can handle missing attributes that can be resolved in grandchild. If there are more complex cases, I think that rule can at least resolve aggregate functions and go back to this rule to complete resolution.

davies replied:
Can we make sure this query work:
```
 select  sum() over () from t where a > 0 group by a having avg(b) > 0 order by avg(c), a 
```

gatorsmile replied:
Let me add a test case to cover it.

gatorsmile replied:
```
select sum(product) over () from windowData where product > 3
group by product having avg(month) > 0
order by avg(area), product
```
This is not resolvable unless we reimplement the rule `ResolveAggregateFunctions`. So far, the rule assumes Sort and Aggregate are adjacent. If you read the plan, you will understand this is not true. This is not a simple change. Do you want me to do it in a separate JIRA?

```
'Sort ['avg('area) ASC,'product ASC], true
+- Project [_c0#61L]
   +- Project [_w0#62L,_c0#61L,_c0#61L]
      +- Window [_w0#62L], [(sum(_w0#62L),mode=Complete,isDistinct=false) windowspecdefinition(ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) AS _c0#61L]
         +- Project [_w0#62L]
            +- Filter havingCondition#63: boolean
               +- Aggregate [product#2], [cast(((avg(cast(month#0 as bigint)),mode=Complete,isDistinct=false) > cast(0 as double)) as boolean) AS havingCondition#63,cast(product#2 as bigint) AS _w0#62L]
                  +- Filter (product#2 > 3)
                     +- Subquery windowdata
                        +- LogicalRDD [month#0,area#1,product#2], MapPartitionsRDD[1] at apply at Transformer.scala:22
```

davies replied:
So this seems that the rule in `ResolveAggregateFunctions` does not really resolve the missing attributes, we could keep that rule unchanged in this PR.

If it's not trivial to fix this, we could create another JIRA for that.

gatorsmile replied:
Unfortunately, it is not trivial. : ( So far, the current rule `ResolveSortReferences` only can handle the missing attributes. In this case, we have to push the aggregate function down to the underlying `Aggregate`. Thus, it does not work.

gatorsmile replied:
@cloud-fan I will let `ResolveAggregateFunctions` handle the missing attribute resolution as long as the child of Sort is Aggregate.
```scala
// Skip sort with aggregate. This will be handled in ResolveAggregateFunctions
case sa @ Sort(_, _, child: Aggregate) => sa
```
When rewriting `ResolveSortReferences` in another PR, I will try to make the behaviors of both rules identical for resolving the missing attributes.

cloud-fan replied:
can we put the problematic query in an ignored test? just in case we don't forget it...

gatorsmile replied:
Sure, I will add it.

davies commented:
LGTM, left two comments.

@marmbrus Could you take another pass on it?

gatorsmile committed:
address comments.

gatorsmile committed:
address comments.

SparkQA commented:
**[Test build #50459 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50459/consoleFull)** for PR 10678 at commit [`ddfebbf`](https://github.com/apache/spark/commit/ddfebbf2bde7a30f68271fbc8ef11705a744f0c8).

SparkQA commented:
**[Test build #50459 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50459/consoleFull)** for PR 10678 at commit [`ddfebbf`](https://github.com/apache/spark/commit/ddfebbf2bde7a30f68271fbc8ef11705a744f0c8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50459/
Test PASSed.

gatorsmile committed:
Added a test case that we need to fix in the next PR.

SparkQA commented:
**[Test build #50479 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50479/consoleFull)** for PR 10678 at commit [`c2964da`](https://github.com/apache/spark/commit/c2964daad833f3a9b1c05aa5d082e356a776c70c).

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -521,38 +522,96 @@ class Analyzer(
           s // Nothing we can do here. Return original plan.
+        } else {
+          // Add the missing attributes into projectList of Project/Window or
+          //   aggregateExpressions of Aggregate, if they are in the inputSet
```
nit: too many spaces

cloud-fan commented:
LGTM, pending tests

rxin commented:
Let's get @marmbrus to take a look at this one too.

SparkQA commented:
**[Test build #50479 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50479/consoleFull)** for PR 10678 at commit [`c2964da`](https://github.com/apache/spark/commit/c2964daad833f3a9b1c05aa5d082e356a776c70c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50479/
Test PASSed.

marmbrus commented:
LGTM, thanks for adding a bunch of tests!

Merging to master.

gatorsmile commented:
Thank you everyone! : )

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -521,38 +522,96 @@ class Analyzer(
-      val newOrdering = resolveSortOrders(ordering, grandchild, throws = true)
+        plan: LogicalPlan,
+        child: LogicalPlan): (Seq[SortOrder], Seq[Attribute]) = {
+      val newOrdering = resolveSortOrders(ordering, child, throws = false)
```
Now it won't throw exception for unresolved attribute here, so what will happen if there are unresolved stuff?
